nx/lib/nx/shared.ex

```diff
@@ -407,7 +407,7 @@ defmodule Nx.Shared do
 
     cond do
       backend == Nx.Defn.Expr ->
-        apply(default_impl, args)
+        Nx.Defn.Expr.optional(apply(default_impl, args), function_name, args)
```

Comment:
I would like to check, either here or inside optional, that the result of `apply(default_impl, args)` matches the output shape in `hd(args)`. We should likely check it in the third clause too.